protocols/kad/src/record/store.rs

```diff
@@ -0,0 +1,95 @@
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+mod memory;
+
+pub use memory::MemoryStore;
+
+use crate::K_VALUE;
+use super::*;
+use std::borrow::Cow;
+
+/// The result of an operation on a `RecordStore`.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// The possible errors of a `RecordStore` operation.
+#[derive(Debug)]
+pub enum Error {
+    /// The store is at capacity w.r.t. the total number of stored records.
+    MaxRecords,
+    /// The store is at capacity w.r.t. the total number of stored keys for
+    /// provider records.
+    MaxProvidedKeys,
+    /// The value of a record to be stored is too large.
+    ValueTooLarge,
+}
+
+/// Trait for types implementing a record store.
+///
+/// There are two types of records managed by a `RecordStore`:
+///
+///   1. Regular (value-)records. These records store an arbitrary value
+///      associated with a key which is distributed to the closest nodes
+///      to the key in the Kademlia DHT as per the standard Kademlia "push-model".
+///      These records are subject to re-replication and re-publication as
+///      per the standard Kademlia protocol.
+///
+///   2. Provider records. These records associate the ID of a peer with a key
+///      who can supposedly provide the associated value. These records are
+///      mere "pointers" to the data which may be followed by contacting these
+///      providers to obtain the value. These records are specific to the
+///      libp2p Kademlia specification and realise a "pull-model" for distributed
+///      content. Just like a regular record, a provider record is distributed
+///      to the closest nodes to the key.
+///
+pub trait RecordStore<'a> {
+    type RecordsIter: Iterator<Item = Cow<'a, Record>>;
+    type ProvidedIter: Iterator<Item = Cow<'a, ProviderRecord>>;
+
+    /// Gets a record from the store, given its key.
+    fn get(&'a self, k: &Multihash) -> Option<Cow<Record>>;
+
+    /// Puts a record into the store.
+    fn put(&'a mut self, r: Record) -> Result<()>;
+
+    /// Removes the record with the given key from the store.
+    fn remove(&'a mut self, k: &Multihash);
+
+    /// Gets an iterator over all (value-) records currently stored.
+    fn records(&'a self) -> Self::RecordsIter;
```

Comment:
Is that a good idea? It works for an in-memory records store, but not a records store that stores them on disk.

Reply:
Why do you think it can only be implemented by an in-memory store? That is certainly not my intention, but every record store must provide a means to iterate through all records - that is simply a requirement of the Kademlia protocol.

Reply:
My impression is that the minimum requirement would be an iterator over the record keys/ids. My concerns is that the iterator also contains tons of metadata as well as the record's content, which might not be desirable to keep in memory.

This is a minor concern though.

Reply:
I see - since I implemented the optimisation from the paper that records are skipped from hourly replication if they have recently been received from another peer (i.e. that effectively only 1 of the `k` peers closest to a key replicates a record hourly in most cases) just iterating over the keys may indeed be beneficial in terms of memory use, albeit requiring separate lookups from the storage for every record that is not skipped. We should certainly keep that in mind as a trade-off to (re)consider.